rotkehlchen/chain/evm/decoding/hop/balances.py

```diff
@@ -0,0 +1,150 @@
+import logging
+from collections import defaultdict
+from typing import TYPE_CHECKING
+
+from eth_utils.address import to_checksum_address
+
+from rotkehlchen.accounting.structures.balance import Balance, BalanceSheet
+from rotkehlchen.assets.asset import Asset
+from rotkehlchen.chain.ethereum.interfaces.balances import BalancesSheetType, ProtocolWithBalance
+from rotkehlchen.chain.ethereum.utils import token_normalized_value
+from rotkehlchen.chain.evm.contracts import EvmContract
+from rotkehlchen.chain.evm.decoding.hop.constants import CPT_HOP
+from rotkehlchen.chain.evm.tokens import get_chunk_size_call_order
+from rotkehlchen.constants.misc import ZERO
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.history.events.structures.evm_event import EvmProduct
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTokenKind
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.db.dbhandler import DBHandler
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class HopCommonBalances(ProtocolWithBalance):
+    def __init__(
+            self,
+            database: 'DBHandler',
+            evm_inquirer: 'EvmNodeInquirer',
+    ):
+        super().__init__(
+            database=database,
+            evm_inquirer=evm_inquirer,
+            counterparty=CPT_HOP,
+            deposit_event_types={(HistoryEventType.STAKING, HistoryEventSubType.DEPOSIT_ASSET)},
+        )
+
+    def query_balances(self) -> 'BalancesSheetType':
+        """Queries and returns the balance sheet for staking events.
+
+        This method:
+        1. Retrieves deposit events for staking.
+        2. Groups addresses by staking contract.
+        3. For each contract, queries user balances and rewards.
+        4. Converts balances and rewards to USD and updates the balance sheet."""
+        balances: BalancesSheetType = defaultdict(BalanceSheet)
+        if len(addresses_with_deposits := self.addresses_with_deposits(products=[EvmProduct.STAKING])) == 0:  # noqa: E501
+            return balances
+
+        # Group addresses by staking contract
+        staked_contracts: dict[ChecksumEvmAddress, set[ChecksumEvmAddress]] = defaultdict(set)
+        for user_address, events in addresses_with_deposits.items():
+            for event in events:
+                if event.address:
+                    staked_contracts[event.address].add(user_address)
+
+        for contract_address in staked_contracts:
+            addresses = staked_contracts[contract_address]
+            staking_contract = EvmContract(
+                address=contract_address,
+                abi=self.evm_inquirer.contracts.abi('HOP_STAKING'),
```

Comment:
since the abi is the same don't query it in a loop. Save it in a variable and reuse it. This is a trip to the database each time if I'm not mistaken